### ISSUE TITLE
Spurious instance declarations cause typeclass resolution to backtrack excessively

### ISSUE THREAD

**Issue by hacklex** [FStarLang/FStar#2572]
I'm toying with the recently added typeclass feature... here's my sandbox:
```fst
module Sandbox

module TC = FStar.Tactics.Typeclasses

class equatable (t:Type) = {
  eq: t -> t -> bool;
  reflexivity : (x:t -> Lemma (eq x x));
  symmetry: (x:t -> y:t -> Lemma (requires eq x y) (ensures eq y x));
  transitivity: (x:t -> y:t -> z:t -> Lemma (requires (x `eq` y /\ y `eq` z)) (ensures (x `eq` z)))
} 
instance ( = ) (#t:Type) {|h: equatable t|} = h.eq

class has_mul (t:Type) = { 
  mul : t -> t -> t;
  [@@@TC.no_method] eq: equatable t;
  [@@@TC.no_method] congruence: (x:t -> y:t -> z:t -> w:t 
                                 -> Lemma (requires (x=z) /\ (y=w)) 
                                         (ensures (mul x y)=(mul z w)))
} 

instance ( * ) (#t:Type) {|m: has_mul t|} = m.mul

class has_add (t:Type) = {  
  add : t -> t -> t;
  [@@@TC.no_method] eq: equatable t;
  [@@@TC.no_method] congruence: (x:t -> y:t -> z:t -> w:t 
                                 -> Lemma (requires (x=z) /\ (y=w)) 
                                         (ensures add x y = add z w))  
} 

instance ( + ) (#t:Type) {|a: has_add t|} = a.add

instance int_equatable : equatable int = {
  eq = op_Equality;
  reflexivity = (fun _ -> ());
  symmetry = (fun _ _ -> ());
  transitivity = (fun _ _ _ -> ())
}

instance eq_of_add (t:Type) {| h: has_add t |} : equatable t = h.eq 

instance eq_of_mul (t:Type) {| h: has_mul t |} : equatable t = h.eq 

class semigroup (t:Type) = {  
  [@@@TC.no_method] has_mul : has_mul t; 
  [@@@TC.no_method] associativity: (x:t -> y:t -> z:t -> Lemma (((x * y) * z) = (x * (y * z))))
}

class add_semigroup (t:Type) = {
  [@@@TC.no_method] has_add : has_add t; 
  [@@@TC.no_method] associativity: (x:t -> y:t -> z:t -> Lemma (((x + y) + z) = (x + (y + z))))
}
```

It takes minutes for the last definition to get verified...

**Comment by nikswamy** [comment]
hi @hacklex 

Thanks for this example. I think the problem here is that you've declared the methods `+`, `=`, `*` etc. as instances, when in fact they are not instances of classes. They are just functions. If you change your program replacing `instance ( = )` with `let ( = )`, and likewise for `( * )` and `( + )` then the program checks out very quickly.

I should add a check to detect when the user writes down an illegal instance declaration. In particular, if the result type of an `instance` declaration is not a `class` type, then we should reject it. That would have caught this problem and saved you from stumbling into this weird behavior. 

To give you a sense of what's happening:

At the resolution of `( = )` in the `add_semigroup` lemma, the initial goal is `equatable t`

* We first try `eq_of_mul` and the goal become `has_mul t`

* Then, it turns out that the spurious instance `(+)` has a type ` #t:Type -> {| has_add t |} -> t -> t -> t`, so this actually matches the goal and now the goal becomes `has_add (has_mul t)`. And at this point, this is a nonsensical goal, but marking (+) has an instance allows it to be used to solve any goal. So, we repeatedly try `(+)` backtracking and retrying an exponential number of times.

So, to summarize, (+) etc. are not instances and F* should reject them.
